mqtt_io/server.py

```diff
@@ -1191,16 +1169,48 @@ async def run() -> None:
                 try:
                     await asyncio.gather(*self.critical_tasks)
                 except asyncio.CancelledError:
-                    pass
+                    break
                 except Exception:  # pylint: disable=broad-except
                     _LOG.exception("Exception in critical task:")
             except asyncio.CancelledError:
-                pass
+                break
+            except MQTTException:
+                if counter > 0:
+                    counter -= 1
+                _LOG.exception('Connection to MQTT-Broker failed')
+                if counter == 0:
+                    break
+                await asyncio.sleep(self.config['mqtt'].get('reconnect_delay'))
```

Comment:
Might be best to just set a flag here saying that we'll reconnect or not, so that the `finally` block can run as soon as we're disconnected, clearing the `self.running` event. We should also clear the `self.mqtt_connected` event in the `finally` block too.

Then once we're out of that block, we can sleep if we're reconnecting and then loop back around. (We could also use that flag as the `while` predicate too, so:

```python
reconnect = True
while reconnect:
    try:
        await self._connect_mqtt()
        ...
    except MQTTException:
        _LOG.exception('Connection to MQTT broker failed')
        if counter == 0:
        
            reconnect = False
        counter -= 1
```